### ISSUE TITLE
arXiv '20 | Memory-Efficient Pipeline-Parallel DNN Training

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#206]
https://arxiv.org/pdf/2006.09503.pdf

**Comment by vycezhong** [comment]
Takeaway:
- PipeDream-2BW: 需要2份weight version
- PipeDream-Flush: 只需要1份weight version

基本思路都是gradient accumulation. 每次BW后不急着更新. 

![image](https://user-images.githubusercontent.com/25879526/122889088-11e84080-d375-11eb-9db0-3ad9a78d232b.png)

![image](https://user-images.githubusercontent.com/25879526/122889133-1ad91200-d375-11eb-9e2f-f011ff2a479d.png)

**Comment by vycezhong** [comment]
温故而知新. 

Notations:
- m: the number of micro batches
- d: pipeline depth (the number of partitions)
- B: effective batch size

|methods| #activations | batch size | #weight version | semantics | 
|---|---|---|---|---|
|GPipe #168  | m | B/m | 1| x_{t+1} = x_t - \eta g_t |
|PipeDream #167  | d | B | d | x_{t+1} = x_t - \eta g_{t-d+1} |
|PipeDream-2BW | d | B/m | 2 | x_{t+1} = x_t - \eta g_{t-1} | 
|PipeDream-Flush | d | B/m | 1 | x_{t+1} = x_t - \eta g_t |

**Comment by vycezhong** [comment]
这个Planner也有点意思. 他们考虑了三种情况
- DP + MP
- DP + MP + PP (w/o activation recomputation)
- DP + MP + PP (w activation recomputation)

他们考虑的模型是具有repetitive structures，相同的block重复若干次，比如BERT, GPT都是如此. 所以他们模型可以用一个二维tuple来表示(d, w)，d代表#partition stages, 而w代表#replica. 

另外还需要考虑micro batch size (b)，以及做不做recomputation (r). 

所以总共的decision variables有
- (d, w)
- b: 最大能承受的micro batch size per GPU (只有一个model weight version)
- b': 最大能承受的micro batch size per GPU (有两个model weight version)
- r: 做不做recomputation

他们的metrics是throughput和memory. 目标是在memory下，尽可能增大throughput. 由于搜索空间不大，他们进行了穷举. 算法如下. 

![image](https://user-images.githubusercontent.com/25879526/126306952-e7441838-c1ac-4e7f-b739-42765ab92482.png)

这是搜索出来的一些结果.
![image](https://user-images.githubusercontent.com/25879526/126306979-311e46d1-4e07-4d1c-8fff-c022156b58f9.png)

第二个问题是cost model. 这确实是一个难点. 我觉得做的比较粗糙. 具体参考paper.

**Comment by ENg-122** [comment]
请问下 论文提到的algorithm1中的optimal degree of gradient accumulation是指最优的梯度累积步数吗？
